src/openttd.cpp

```diff
@@ -603,6 +614,42 @@ int openttd_main(int argc, char *argv[])
 		case 'P':
 			scanner->join_company_password = mgo.opt;
 			break;
+		case 'U': {
+			char* p = mgo.opt;
+
+			const char* protocol = "openttd://";
+			if (strncmp(p, protocol, strlen(protocol)) != 0) {
+				fprintf(stderr, "Unknown protocol\n");
+				break;
+			}
+
+			p += strlen(protocol);
+
+			char* network_conn = p;
+			p = strchr(p, '?');
+			*p = '\0';
+			p++;
+			scanner->network_conn = network_conn;
+
+			const char* name = nullptr;
+			const char* value = nullptr;
+			while (ParseQueryString(&name, &value, &p)) {
+				if (strcmp(name, "password") == 0) {
+					scanner->join_server_password = value;
+				} else if (strcmp(name, "join") == 0) {
+					scanner->join_company = value;
+				} else if (strcmp(name, "companypassword") == 0) {
+					scanner->join_company_password = value;
+				} else if (strcmp(name, "version") == 0) {
+					/* Nothing for now */
```

Comment:
I think having multiple versions handling in from the beginning would be ideal. Have a per-platform function for launching a different installed version (by version string) that as a fallback just does nothing.
By having fully functioning multiple versions handling in from the beginning, any installed version can be registered as the primary URL handler, and will always be able to re-launch any other installed version.

Reply:
Hey @nielsmh, I'm wondering how the program would look for other versions that are installed? I tried installing 2 versions of OpenTTD using the Windows installers but only the latest version is kept. This leads me to think that each version would have to be installed in different user-chosen locations...

Reply:
Yes it would have to be a new thing done by something. I don't have a bulletproof idea yet, but something with versions of the game (that support URL handler) registering itself when it starts. So if you have a bunch of different portable versions with this feature, the first time you run one it registers itself with version and path, and you end up with all of them registered and available for URL handlers.

Perhaps also a mode where earlier versions can be registered so an URL-handler-supporting version can convert the URL into the older commandline parameters format and start those.

Reply:
Hmm, yeah, i see what you mean. What information would we need to store? Is it just version number and filepath for each executable? Presumably this could be stored in a simple config file, ignoring weird characters (eg a newline) in the file path. I guess binary file could be used instead to allow for this. 

Maybe for registering old versions, a -r (--register) option could be added that would take the path of an openttd executable and the new version would find its version number and register it?

Also, there would need to be some way to tell if the older cli parameters would need to be used. Could that just be a hardcode `if version < some_version: use old cli` or is that too rigid/fragile?

Reply:
On Windows I would do it in registry.
Key name `HKCU\Software\OpenTTD\Versions\<version>`
Default value = (string) path to exe file
Value `NetworkJoinMode`= (dword) 0 for classic, 1 for -U parameter support

The `<version>` in the key name ("folder") should be the network version of the build, which should usually be the same as the product version string in the exe file's version resource. This means the main URL handler will be able to verify that the exe file pointed to is the actual version intended before launching it, and abort if it's not a match. That would work for even very old versions.

As for registering old versions/versions not supporting the URL handler, it could either be done in a manual way, or an automatic-ish scanner. Manual would be prompting the user to find the exe file for a specific version when they try to join a server where they don't have the version for, or as you suggest having a call to register a specific version. Automatic scanner would probably be best as a separate downloadable tool.

On Unix'y systems, I'd put the database in a similar format (maybe an INI format) in the user's profile `~/.openttd/`.
On macOS, the right way might involve a PList of some kind.

Reply:
Ok, cool.

I don't actually have access to a macOS machine (nor have I even heard of PLists!) but I can give it a go anyway. 

A prompt does make sense. And i suppose the automatic scanner idea can be made at a later date. 

I'll have a look into all this whenever i get some time

Should i just continue in this pr or should i make a pr on my fork to (temporarily) separate the changes? I've never added this much in a pr before...